File: TuringGame/TuringGame/src/main/java/Util/Observable.java
package Util;

/**
 * An Observable object in the Observer/Observable pattern.
 * Maintains a list of observer and notifies them
 * whenever an internal state changed.
 *
 * @author Frédéric Servais
 */
public interface Observable {

    /**
     * Adds an observer to the list of observers.
     *
     * @param observer The observer to be added.
     */
    void addObserver(Observer observer);

    /**
     * Removes the observer from the list.
     *
     * @param observer The  observer to be removed.
     */
    void removeObserver(Observer observer);

    /**
     * Notifies all observers by calling their 'update' method.
     */
    void notifyObservers();
}

File: Tetris/Tetris/TetrisGui/util/Observable.cpp
#include "../util/Observable.h"
#include <algorithm>

void Observable::addObserver(std::shared_ptr<Observer> observer) {
    observers.push_back(observer);
}

void Observable::removeObserver(std::shared_ptr<Observer> observer) {
    auto it = std::find(observers.begin(), observers.end(), observer);
    if (it != observers.end()) {
        observers.erase(it);
    }
}

void Observable::notify(GameState state) const {

    for (std::shared_ptr<Observer>  observer : observers) {
        observer->update(state);
    }
}
